src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -2498,6 +2499,337 @@ public static decimal Average<TSource>(this IEnumerable<TSource> source, Func<TS
         {
             return Enumerable.Average(Enumerable.Select(source, selector));
         }
+
+        public static TSource MaxBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector)
+        {
+            return source.MaxBy(keySelector, null);
+        }
+
+        public static TSource MaxBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IComparer<TKey> comparer)
+        {
+            if (source == null) throw Error.ArgumentNull("source");
+            if (keySelector == null) throw Error.ArgumentNull("keySelector");
+            if (comparer == null) comparer = Comparer<TKey>.Default;
+            TSource value = default(TSource);
+            if (value == null)
+            {
+                using (IEnumerator<TSource> e = source.GetEnumerator())
+                {
+                    do
```

Comment:
On the next line an empty sequence would cause `null` to be immediately returned, which is the correct behaviour for an empty sequence of a nullable type, to be consistent with e.g. `new string[0].Max()` and `new DateTime?[0].Max()`, etc. which return `null`. Similarly the code path for non-nullable types throw an `InvalidOperationException` to be consistent with out `new DateTime[0].Max()` throws, etc.